File: 2026-09-22-pb158-lecture-3-finishing-io-and-advanced-io/demo/close/close_twice.c
/* close_twice.c — CloseHandle a handle, then CloseHandle it again. Windows, in C.
 *
 *   close_twice.exe
 *
 * WHAT THIS SHOWS
 *   The Windows twin of close_twice.py. The second CloseHandle on the same
 *   handle value fails with ERROR_INVALID_HANDLE (6): the first call released
 *   it, and the value may already belong to something else in this process.
 *   Microsoft's documentation adds a twist worth seeing live: "If the
 *   application is running under a debugger, the function will throw an
 *   exception if it receives either a handle value that is not valid ... This
 *   can happen if you close a handle twice." NtTrace64 IS a debugger — so under
 *   nmake trace the second close does not fail quietly, it raises.
 *
 * WHAT TO LOOK AT
 *   Untraced: "second CloseHandle: failed, error 6". Traced: NtClose twice, the
 *   second returning STATUS_INVALID_HANDLE (0xC0000008), and the debugger
 *   exception NtTrace reports.
 *
 * WHY IT MATTERS
 *   Same rule as Linux: a failed or repeated close is a bug report, not
 *   something to retry. "An application should call CloseHandle once for each
 *   handle it opens."
 *
 * Build: nmake /f Makefile.msvc     Trace: nmake /f Makefile.msvc trace
 */
#include <windows.h>
#include <stdio.h>
int main(void) {
    HANDLE h = CreateFileA("closed.txt", GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
    if (h == INVALID_HANDLE_VALUE) { fprintf(stderr, "CreateFile: %lu\n", GetLastError()); return 1; }
    DWORD n; WriteFile(h, "one\ntwo\nthree\n", 14, &n, NULL);
    printf("first  CloseHandle: %s\n", CloseHandle(h) ? "ok" : "failed");      /* the handle value is released HERE */
    if (CloseHandle(h)) printf("second CloseHandle: ok (unexpected)\n");        /* "the retry" */
    else printf("second CloseHandle: failed, error %lu (ERROR_INVALID_HANDLE is 6)\n", GetLastError());
    return 0;
}

File: 2026-09-22-pb158-lecture-3-finishing-io-and-advanced-io/demo/close/close_twice.py
#!/usr/bin/env python3
"""close_twice.py — close a descriptor, then close it again. What does the second close hit?

    python3 close_twice.py            (Linux, macOS, Windows)

WHAT THIS SHOWS
  The carried Lecture 2 slide says: never retry a failed close(). This program
  shows the mechanical reason. It opens a file, closes it, and closes the SAME
  descriptor number again — which is exactly what a retry does. The second
  close fails with EBADF (bad file descriptor): the number was released by the
  first close, whether or not that first close reported an error. On Linux the
  manual page is explicit: the kernel "always releases the file descriptor
  early in the close operation, freeing it for reuse", so a retry can only ever
  hit nothing — or, in a program with other threads, SOMEBODY ELSE's newly
  opened file that was handed the same number.

WHAT TO LOOK AT
  Under strace (make trace): close(3) = 0, then close(3) = -1 EBADF. Two
  crossings, one descriptor, one of them already gone.

WHY IT MATTERS
  When a call's failure leaves you unable to tell what state you are in, the
  only safe move is to report it and stop retrying. Three storage engines
  reached the same conclusion for fsync in 2018 (Part 5 / Part 6).
"""
import os, sys

fd = os.open("closed.txt", os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o644)
os.write(fd, b"one\ntwo\nthree\n")
os.close(fd)                       # the descriptor number is released HERE, error or not
print(f"first  close({fd}): ok")
try:
    os.close(fd)                   # "the retry": the number no longer refers to anything of ours
    print(f"second close({fd}): ok (unexpected)")
except OSError as e:
    print(f"second close({fd}): {e.errno} {e.strerror}")   # EBADF on Linux and Windows alike
